crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,961 @@
+use tokio::{
+    net::TcpListener,
+    sync::{
+        broadcast::{self, Receiver as BroadcastReceiver, Sender as BroadcastSender},
+        oneshot::{self, Sender as OneshotSender},
+    },
+    task,
+};
+
+/// The default port that Bevy will listen on.
+///
+/// This value was chosen randomly.
+pub const DEFAULT_PORT: u16 = 15702;
+
+const CHANNEL_SIZE: usize = 16;
+
+/// Add this plugin to your [`App`] to allow remote connections to inspect and modify entities.
+///
+/// By default, this is [`DEFAULT_PORT`]: 15702.
+pub struct RemotePlugin {
+    /// The port that Bevy will listen on.
+    pub port: u16,
+}
+
+/// A resource containing the port number that Bevy will listen on.
+#[derive(Resource, Reflect)]
+pub struct RemotePort(pub u16);
+
+/// The type of a function that implements a remote verb (`GET`, `QUERY`, etc.)
+///
+/// The first parameter is the JSON value of the `params`. Typically, an
+/// implementation will deserialize these as the first thing they do.
+///
+/// The returned JSON value will be returned as the response. Bevy will
+/// automatically populate the `status` and `id` fields before sending.
+pub type RemoteVerb =
+    for<'w, 'a> fn(Value, &'w mut World, &'a AppTypeRegistry) -> AnyhowResult<Value>;
+
+/// Holds all implementations of verbs known to the server.
+///
+/// You can add your own custom verbs to this list.
+#[derive(Resource, Deref, DerefMut)]
+pub struct RemoteVerbs(pub HashMap<String, RemoteVerb>);
+
+/// A single request from a Bevy Remote Protocol client to the server,
+/// serialized in JSON.
+///
+/// The JSON payload is expected to look like this:
+///
+/// ```json
+/// {
+///     "request": "GET",
+///     "id": 0,
+///     "params": {
+///         "data": {
+///             "entity": 4294967298,
+///             "components": [
+///                 "bevy_transform::components::transform::Transform"
+///             ]
+///         }
+///     }
+/// }
+/// ```
+#[derive(Serialize, Deserialize, Clone)]
+pub struct BrpRequest {
+    /// The verb: i.e. the action to be performed.
+    pub request: String,
+
+    /// Arbitrary data that will be returned verbatim to the client as part of
+    /// the response.
+    pub id: Value,
+
+    /// The parameters, specific to each verb.
+    ///
+    /// These are passed as the first argument to the verb handler.
+    pub params: Value,
+}
+
+/// `GET`: Retrieves one or more components from the entity with the given
+/// ID.
+///
+/// The server responds with a `BrpResponse::Get`.
+#[derive(Serialize, Deserialize, Clone)]
+pub struct BrpGetRequest {
+    /// The ID of the entity from which components are to be requested.
+    pub entity: Entity,
+
+    /// The *full paths* of the component types that are to be requested
+    /// from the entity.
+    ///
+    /// Note that these strings must consist of the *full* type paths: e.g.
+    /// `bevy_transform::components::transform::Transform`, not just
+    /// `Transform`.
+    pub components: Vec<String>,
+}
```

Comment:
This is a more subjective comment, but now that request handlers are extensible and modular, I think it would be nice to move these default ones into their own file, e.g., `core_handlers.rs`.

Reply:
They're now in `builtin_verbs.rs`.